src/ApplicationConfigInjectionTrait.php

```diff
@@ -0,0 +1,306 @@
+                        'Allowed HTTP methods for a route must be in form of an array; received "%s"',
+                        gettype($methods)
+                    ));
+                }
+            } else {
+                $methods = Route::HTTP_METHOD_ANY;
+            }
+            $name    = isset($spec['name']) ? $spec['name'] : null;
+            $route   = new Route($spec['path'], $spec['middleware'], $methods, $name);
+
+            if (isset($spec['options'])) {
+                $options = $spec['options'];
+                if (! is_array($options)) {
+                    throw new Container\Exception\InvalidArgumentException(sprintf(
+                        'Route options must be an array; received "%s"',
+                        gettype($options)
+                    ));
+                }
+
+                $route->setOptions($options);
+            }
+
+            $this->route($route);
+        }
+    }
+
+    /**
+     * Create and return the pipeline map callback.
+     *
+     * The returned callback has the signature:
+     *
+     * <code>
+     * function ($item) : callable|string
+     * </code>
+     *
+     * It is suitable for mapping pipeline middleware representing the application
+     * routing o dispatching middleware to a callable; if the provided item does not
+     * match either, the item is returned verbatim.
+     *
+     * @return callable
+     */
+    private function createPipelineMapper()
+    {
+        return function ($item) {
+            if ($item === Application::ROUTING_MIDDLEWARE) {
+                return [$this, 'routeMiddleware'];
+            }
+
+            if ($item === Application::DISPATCH_MIDDLEWARE) {
+                return [$this, 'dispatchMiddleware'];
+            }
+
+            return $item;
+        };
+    }
+
+    /**
+     * Create the collection mapping function.
+     *
+     * Returns a callable with the following signature:
+     *
+     * <code>
+     * function (array|string $item) : array
+     * </code>
+     *
+     * When it encounters one of the self::*_MIDDLEWARE constants, it passes
+     * the value to the `createPipelineMapper()` callback to create a spec
+     * that uses the return value as pipeline middleware.
+     *
+     * If the 'middleware' value is an array, it uses the `createPipelineMapper()`
+     * callback as an array mapper in order to ensure the self::*_MIDDLEWARE
+     * are injected correctly.
+     *
+     * If the 'middleware' value is missing, or not viable as middleware, it
+     * raises an exception, to ensure the pipeline is built correctly.
+     *
+     * @return callable
+     */
+    private function createCollectionMapper()
+    {
+        $pipelineMap = $this->createPipelineMapper();
+        $appMiddlewares = [
+            Application::ROUTING_MIDDLEWARE,
+            Application::DISPATCH_MIDDLEWARE,
+        ];
+
+        return function ($item) use ($pipelineMap, $appMiddlewares) {
```

Comment:
Can't be static, as it's closing over provided values. Additionally, I don't see what value that would add; this is returned from a private method, so those extending would not be able to re-bind anyways.